templates/todo/projects/csharp-sql/infra/main.parameters.json

```diff
@@ -7,6 +7,15 @@
     },
     "location": {
       "value": "${AZURE_LOCATION}"
+    },
+    "principalId": {
+      "value": "${AZURE_PRINCIPAL_ID}"
+    },
+    "sqlAdminPassword" : {
+      "value": "$(randomPassword)"
+    },
+    "appUserPassword" : {
+      "value": "$(randomPassword)"
```

Comment:
Since we generate a new random password each time this file is processed - does this mean that every `azd provision` ends up resetting admin and app password, because the deployment script runs every time?

Reply:
This is unfortunately true.

Reply:
The super-crazy thing we could do here is change the `randomPassword` invocation into `existing-keyvault-secret-or-random-password-fallback` (not actual name, just to capture the semantics). The way it would work is, when we "plan" the deployment, we would look to see if the keyvault with the secret is already provisioned, and if yes, grab the secret and use it as the password value. Otherwise generate new random password. This way the same value would be used during re-provisioning. 

I believe we have all the inputs to do this, or at least the user could give us a hint as to which KeyVault to look for--that bit could be passed as invocation parameter to `randomPassword`.

Reply:
If we want to pursue the above idea, I propose we open a new issue to do it as a separate PR.

Reply:
I do think we need to figure something out here - I am worried that as is we are going to run into weird issues like running `azd provision` again breaks the application that is deployed because the user password changes and the deployed app is now using incorrect credentials. What you outline here seems like a reasonable thing to explore.  I do think there's an interesting question as to how the key-vault we store the secret in gets provisioned.

One other possibility (maybe?) would be to keep the behavior we have, but try to hack things such that we use the parameter for the initial deployment, but then set some output in the bicep, and use the existence of that output in the environment as a way to know "oh the initial deploy already happened".  It feels like that trick, paired with the the `getSecret` function in bicep and conditional deployment might be enough power to dig ourselves out of this hole.

Reply:
The app will work because it always gets the secret value from KV at startup. Worst case it might need to be restarted after infra re-provisioning, but I think that is reasonable, and probably necessary anyway.

I guess the question is: do you want to hold this PR off while we solve the problem that DB provisioning is "less incremental" that it could/should be? My vote would be to go ahead with the PR as-is and fix the re-provisioning as a separate change, because I think there is enough developer value in what we have here, but I can be convinced otherwise.

Reply:
I like the idea of getting what we have in there and fixing this after.

Reply:
(will discuss offline with Matt)

Reply:
Implemented the `secretOrRandomPassword` idea, please take another look @ellismg  

I need to fix a couple of tests and test this implementation, but what you can see should give you a good idea about how it is supposed to work.